janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java

```diff
@@ -697,31 +698,79 @@ private void checkConnectionConstraintOrCreateConnectionConstraint(JanusGraphVer
+        }
+        if (multiplicity.isUnique(Direction.IN)) {
+            if (!Iterables.isEmpty(query(inVertex).type(label).direction(Direction.IN).edges()))
+                    throw new SchemaViolationException("An edge with the given label already exists on the in-vertex and the label [%s] is in-unique", label.name());
+        }
+    }
+
+    public JanusGraphEdge addEdge(String relationId, JanusGraphVertex outVertex, JanusGraphVertex inVertex, EdgeLabel label){
         verifyWriteAccess(outVertex, inVertex);
         outVertex = ((InternalVertex) outVertex).it();
         inVertex = ((InternalVertex) inVertex).it();
         Preconditions.checkNotNull(label);
         checkConnectionConstraintOrCreateConnectionConstraint(outVertex, inVertex, label);
         Multiplicity multiplicity = label.multiplicity();
         TransactionLock uniqueLock = getUniquenessLock(outVertex, (InternalRelationType) label,inVertex);
+        if (!graph.getConfiguration().allowEdgeIdSetting()) {
+            throw new UnsupportedOperationException(String.format("Provided relation id [%s] is ignored because edge id setting is not enabled", relationId));
+            // relationId = null;
+        }
         uniqueLock.lock(LOCK_TIMEOUT);
         try {
             //Check uniqueness
             if (config.hasVerifyUniqueness()) {
-                if (multiplicity==Multiplicity.SIMPLE) {
-                    if (!Iterables.isEmpty(query(outVertex).type(label).direction(Direction.OUT).adjacent(inVertex).edges()))
-                            throw new SchemaViolationException("An edge with the given label already exists between the pair of vertices and the label [%s] is simple", label.name());
-                }
-                if (multiplicity.isUnique(Direction.OUT)) {
-                    if (!Iterables.isEmpty(query(outVertex).type(label).direction(Direction.OUT).edges()))
-                            throw new SchemaViolationException("An edge with the given label already exists on the out-vertex and the label [%s] is out-unique", label.name());
-                }
-                if (multiplicity.isUnique(Direction.IN)) {
-                    if (!Iterables.isEmpty(query(inVertex).type(label).direction(Direction.IN).edges()))
-                            throw new SchemaViolationException("An edge with the given label already exists on the in-vertex and the label [%s] is in-unique", label.name());
+                checkUniqueness(multiplicity, outVertex, inVertex, label);
+            }
+            StandardEdge edge = new StandardEdge(IDManager.getTemporaryRelationID(temporaryIds.nextID()), label, (InternalVertex) outVertex, (InternalVertex) inVertex, ElementLifeCycle.New);
+            if (config.hasAssignIDsImmediately()) {
+                if(relationId!=null){
+                    long long_relationId = LongEncoding.decode(relationId);
```

Comment:
Long overflow can happen if the given string is long enough. But anyway, I think there is some required pattern for auto-generated edge id, though I am not sure the reason behind that. Maybe someone familiar with that can answer.

Reply:
When the given string is very long, the decode long value is negative, which would throw `java.lang.IllegalArgumentException: Expected non-negative number: -2278243806334820789`. Also, in the vertex part, the vertexid has some limitations, like non-negative value and [IDManager check](https://github.com/9lan9/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java#L509). So it should be the user's responsibility to make sure the given id is valid.

Reply:
Why not let users provide a long edge ID, similar to custom vertex ID?